[zenithyap] iP

zenithyap opened the pull request:
# Pix User Guide :poop: 
> “Your mind is for having ideas, not holding them.” – David Allen  

  **Pix helps to remember stuff for you so that you do not have to! It is a Command-Line-Interface based task manager that is ~~easy~~ super easy to learn and quick to use.**

- [Installation](#installation)
- [Features](#features)
- [Commands](#commands)

## Installation
1. Ensure that you have java `11` installed.
2. Download the [ip.jar](https://github.com/zenithyap/ip/releases/download/Level-9/ip.jar) file.
3. Open a terminal and navigate to the folder the file is downloaded in.
4. Run the command `java -jar ip.jar`
5. You should see the following:
 ```
 Hello, this is
 ______    __  ___   ___ 
 |   _  \  |  | \  \ /  / 
 |  |_)  | |  |  \  V  /  
 |   ___/  |  |   >   <   
 |  |      |  |  /  .  \  
 | _|      |__| /__/ \__\

 --------------------------------------------------------------
  How can I assist you?
 --------------------------------------------------------------
 ```

## Features
### Manages Your Tasks
- [X] Able to add/delete tasks
- [X] Supports 3 tasks, todo, deadlines, and events
- [X] Mark/Unmark tasks 
- [X] Find tasks by keyword

## Commands
### `todo`  
Adds a todo task.  
  
Format: `todo <Description>`

### `dline`  
Adds a deadline task. By to be specified as yyyy-MM-dd HH:mm.
  
Format: `dline <Description> / <By>`

### `event`  
Adds an Event task. From and To to be specified as yyyy-MM-dd HH:mm.
  
Format: `event <Description> / <From> / <To>`

### `ls`  
Lists all the tasks.
  
Format: `ls`

### `lsd`  
Lists deadlines that are due by the date and events that area ongoing during the date. Date format as yyyy-MM-dd 
  
Format: `lsd <Date>`

### `mk`  
Marks a task as done.
  
Format: `mk <Index>`

### `unmk`  
Unmarks a task.
  
Format: `unmk <Index>`

### `rm`  
Removes a task from the list.
  
Format: `rm <Index>`

### `find`  
Find tasks that contains a keyword.
  
Format: `find <Keyword>`

### `bye`  
Exits pix.
  
Format: `bye`
  
# Pix's main code:
```Java
public class Pix {
    /**
     * Main function which runs Pix's logic.
     *
     * @param args Arguments provided.
     */
    public static void main(String[] args) {
        MyData data = new MyData();
        Ui ui = new Ui();
        try {
            data.loadData();
        } catch (IOException e) {
            System.out.println(e.getMessage());
        }
        Parser parser = new Parser(data);
        Ui.display();

        Scanner sc = new Scanner(System.in);

        while (sc.hasNextLine()) {
            String command = sc.nextLine();
            try {
                Command parsed = parser.parse(command);
                parsed.execute(data, ui);
            } catch (PixException e) {
                System.out.println(e.getMessage());
            }
            if (command.equals("bye")) {
                break;
            }
        }
        sc.close();
    }
}
```

Commit:
Add Gradle support

zenithyap committed:
added greetings, echo and exit functionality

zenithyap committed:
Add ability to store and display text

zenithyap committed:
Added ability to mark tasks as done + revised OOP structure

zenithyap committed:
Added support for 3 tasks : todos, deadlines and events

zenithyap committed:
Use the input/output redirection technique to semi-automate the testing of Duke

zenithyap committed:
Added error handling for incorrect inputs

zenithyap committed:
Added support to delete tasks from list

Commit:
Add save functionality

zenithyap committed:
Add dates and times

zenithyap committed:
Add times and dates

zenithyap committed:
Add more OOP

zenithyap committed:
Organize into packages

zenithyap committed:
Organize into packages

zenithyap committed:
Merge remote-tracking branch 'origin/add-gradle-support'

zenithyap committed:
Add gradle

Commit:
Add some JUnit tests

Commit:
Package App as JAR file

Commit:
Package App into JAR file

Commit:
Add java docs

Commit:
Follow coding standard

Commit:
Add function to search task by keywork

Commit:
Merge branch-A-JavaDoc and branch branch-A-CodingStandard

Commit:
Merge branch-Level-9

Commit:
Merge branch-Level-9

zenithyap committed:
Update README.md

Commit:
Add checkstyle

Commit:
Add checkstyle

zenithyap committed:
Update README.md

zenithyap committed:
Update README.md

zenithyap committed:
Add checkstyle

zenithyap committed:
Merge branch 'master' of https://github.com/zenithyap/ip

jerrrren commented on src/main/java/pix/commands/AddDeadline.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * AddDeadline class to add deadline to list of tasks.
+ */
+public class AddDeadline extends Command {
```
Class name should be noun so wouldn't DeadlineCommand would be better?

jerrrren commented on src/main/java/pix/commands/Bye.java:
```diff
@@ -0,0 +1,13 @@
+/**
+ * Bye class which exits pix.
+ */
+public class Bye extends Command {
```
Class name should be noun so wouldn't ByeCommand would be better?

jerrrren commented on src/main/java/pix/data/MyData.java:
```diff
@@ -0,0 +1,136 @@
+ */
+public class MyData {
+    /** Arraylist to keep track of all tasks. */
+    private final ArrayList<Task> data = new ArrayList<Task>();
```
Is it more appropriate to name this variable as tasks since it is an arraylist consisting of tasks which is the plural of task?

jerrrren commented on src/main/java/pix/data/MyData.java:
```diff
@@ -0,0 +1,136 @@
+     * Saves the list of tasks to a text file on local device.
+     */
+    public void saveToFile() {
+        String home = System.getProperty("user.dir");
```
Variable name can be more specific instead of something generic like home, possibly name it like homeDirectory?

jerrrren commented on src/main/java/pix/tasks/Deadline.java:
```diff
@@ -0,0 +1,43 @@
+ */
+public class Deadline extends Task {
+    /** Date and time of deadline. */
+    protected LocalDateTime by;
```
If possible try to replace this variable with another noun?

jerrrren reviewed (commented):
In general, good job in ensuring that formatting follows the java standard. However all the command subclasses are not named in a noun fashion so maybe it would be better for you to use an alternative class name.

zenithyap committed:
Add GUI

zenithyap committed:
Add GUI

Commit:
Build Pix Jar file

Changed the build.gradle file to build the Pix.jar file

hhchinh2002 commented on src/main/java/pix/commands/AddEvent.java:
```diff
@@ -0,0 +1,37 @@
+/**
+ * AddEvent class to add event to list of tasks.
+ */
+public class AddEvent extends Command {
```
Would naming classes representing different types of tasks and event as "AddSomething" lower the class dealing scopes? E.g In the future, the class may need to encapsulate some other information about the task, the class name may sound confused in that case since it suppose to use as adding event/task only.

hhchinh2002 commented on src/main/java/pix/parser/Parser.java:
```diff
@@ -0,0 +1,260 @@
+     * @throws PixException If no description, due date or invalid format of date given.
+     */
+    public Command parseDeadline(String[] slashed) throws PixException {
+        if (slashed.length != 2 || removeCommand(slashed[0]).isEmpty() || removeCommand(slashed[1]).isEmpty()) {
```
Would it be more readable if these conditions is  initialized as seperated boolean variables instead invoking methods inside if statement?

hhchinh2002 commented on src/main/java/pix/parser/Parser.java:
```diff
@@ -0,0 +1,260 @@
+                    + "e.g. deadline 'description' / 'yyyy-MM-dd HH-mm'");
+        }
+        try {
+            return new AddDeadline(removeCommand(slashed[0]), removeCommand(slashed[1]));
```
Same as above comment, since removeCommand(Slashed[0]) and removeCommand(Slashed[1]) is referred to twice, would it be neater if we initialize these 2 objects with a more readable name and calling to it when needed?

hhchinh2002 reviewed (commented):
Overall, the codes are great, just some small suggestions to look more readable, I learned a lot from doing reviews on your code,

Commit:
Add assertions

Added assert feature (not JUnit assertions) to document important assumptions that should hold at various points in the code.

zenithyap committed:
Refine code quality

Critically examined the code and refactor to improve the code quality where necessary, based on 'Code Quality' topics learnt so far.

zenithyap committed:
Merge pull request #1 from zenithyap/branch-A-Assertions

Add assertions

zenithyap committed:
Refine code quality

zenithyap committed:
Merge branch 'master' into branch-A-CodeQuality

zenithyap committed:
Merge pull request #2 from zenithyap/branch-A-CodeQuality

Branch a code quality

zenithyap committed:
Revert "Branch a code quality"

zenithyap committed:
Merge pull request #3 from zenithyap/revert-2-branch-A-CodeQuality

Revert "Branch a code quality"

zenithyap committed:
Fix

zenithyap committed:
Revert main to d347412

zenithyap committed:
Refine code quality

Critically examined the code and refactor to improve the code quality where necessary, based on the 'Code Quality' topics taught in CS2103T.

zenithyap committed:
Add help command

zenithyap committed:
Fix naming conventions

zenithyap committed:
Fix event and dline commands error

zenithyap committed:
Add user guide

zenithyap committed:
Update user guide

zenithyap committed:
Update user guide

zenithyap committed:
Resize image in user guide

zenithyap committed:
Update user guide

zenithyap committed:
Update README.md

zenithyap committed:
Fix more bugs

zenithyap committed:
Merge branch 'master' of https://github.com/zenithyap/ip

zenithyap committed:
Fix more bugs
